nexus/src/app/sled.rs

```diff
@@ -231,4 +233,166 @@ impl super::Nexus {
         self.db_datastore.service_upsert(opctx, service).await?;
         Ok(())
     }
+
+    // OPTE V2P mappings
+
+    /// Ensure that the necessary v2p mappings for an instance are created
+    pub async fn create_instance_v2p_mappings(
```

Comment:
comment added in 3d5e677d

Reply:
New comment LGTM, except that I don't quite follow why this is a problem if the caller targets a subset of sleds but not the complete set of sleds. (The race I had in mind was one where one call starts updating to the correct mappings for time T; that call gets swapped out; the instance moves, generating new mappings for T+1; a second call applies the T+1 mappings; and then the first call resumes and applies the now-obsolete mappings from T. If that's possible, I don't think it matters whether the mappings involve all the sleds or just some of them.)

Reply:
This is the scenario you're describing (I think): with four sleds, where two instances (each with one NIC) that share a VPC are allocated to sleds A and B:

```
|A| |B| |C| |D|
|-| |-| |-| |-|
|1| |2| | | | |
| | | | | | | |
| | | | | | | |
```

The minimum number of OPTE V2P mappings required (in addition to the ones automatically created) for these two instances to communicate is:

```
on sled A: [
  instance 1 -> sled A (automatically created by OPTE)
  instance 2 -> sled B
]
on sled B: [
  instance 1 -> sled A
  instance 2 -> sled B (automatically created by OPTE)
]
on sled C: [
]
on sled D: [
]
```

If an additional instance 3 in the same VPC is allocated to sled C, then the changes required are:

```
- add a mapping on sled A: instance 3 -> sled C
- add a mapping on sled B: instance 3 -> sled C
- add a mapping on sled C: instance 1 -> sled A
- *add a mapping on sled C: instance 2 -> sled B*
```

If instance 2 is migrated from B to D, then the changes required *before instance 3 exists* are:

```
- delete the mapping on sled A: instance 2 -> sled B
- add a mapping on sled A: instance 2 -> sled D
```

If instance 3 exists, then additional changes are required:

```
- delete the mapping on sled C: instance 2 -> sled B
- *add a mapping on sled C: instance 2 -> sled D*
```

The mappings surrounded in asterisks conflict with each other. If the B->D migration occurs in the middle of the instance C creation saga, it's possible the `sled C: instance 2 -> sled B` mapping will be applied last, which is incorrect.

If there are more unnecessary mappings, then I think you're right, it doesn't help. I'm not sure what I was thinking.

Reply:
I was also unsure of OPTE's behaviour: if `instance 2 -> sled B` and `instance 2 -> sled D` are applied to the same sled, what happens? I tried

```
opteadm set-v2p 172.30.0.5 A8:40:25:F2:D4:4C fd00:1122:3344:100::1 912
opteadm set-v2p 172.30.0.5 A8:40:25:F2:D4:4C fd00:1122:3344:101::1 912
```

and dump-v2p showed:

```
IPv4 mappings
----------------------------------------------------------------------
VPC IP                   VPC MAC ADDR      UNDERLAY IP
172.30.0.5               A8:40:25:F2:D4:4C fd00:1122:3344:101::1
```

so it appears that the last mapping wins (not that two mappings would exist or that an error would be thrown).

Reply:
I thought about this some more and I think your original comment was right. tl;dr: my kingdom for reliable persistent workflows. More analysis below.

---

> If there are more unnecessary mappings, then I think you're right, it doesn't help. I'm not sure what I was thinking.

I might have been mistaken. I think it *might* help, albeit kind of indirectly. Here's the scenario I think you're describing:

- we start with instance 1 on sled A
- we create instance 2 on sled B
- as part of that saga, we read the locations of the other instances in the VPC and see instance 1 is on sled A. So we know we need to create two mappings:
  - tell sled A "instance 2 is on sled B"
  - tell sled B "instance 1 is on sled A"
- then, while this saga is still in progress, we migrate instance 1 to sled C.

We want to end up with the following knowledge on each sled (ignoring the identity mappings):

- Sled A: no longer in use
- Sled B: instance 1 is on sled C
- Sled C: instance 2 is on sled B

And we have two problems:

1. Sled B might get the wrong information about instance 1: instance 1's migration-completion thread and instance 2's creation thread are racing to tell sled B where instance 1 is (instance 2's creator thinks it's sled A, the migration thinks it's sled C).
2. Both sleds A and C will be told where instance 2 is (the end-of-migration thread will tell C and instance 2's creation thread will tell A). This creates an unnecessary mapping, but not an incorrect one.

The trick is that if, when we created instance 1, we told *all* the other sleds about it, then when we create instance 2, there is no need for that creation saga to tell sled B where to find instance 1--we know that that was done already. So the instance creator doesn't have to do anything that will conflict with the migration, which prevents the race.

---

Note that I think this only works because the instance creation saga guarantees that all the V2P mappings exist before the instance starts to run. If we did it in the other order, we could have this race instead:

- Instance 1 is created and starts running on sled A
- Afterward, we start telling other sleds that I1 is on sled A
- But the instance is created, so it can migrate! A migration moves I1 to sled B
- After migration, the end-of-migration thread starts telling other sleds that I1 is on sled B

*This* is the race I think I had in mind--you have two broadcasters of an instance's location who have different ideas about where the instance is. This isn't a problem for creation with the saga ordered as it is today, but I think it *is* a problem for an instance that migrates once and then migrates again before we finish dealing with the consequences of the first migration (at least if we defer programming V2P mappings until after migration is done).

Reply:
> The trick is that if, when we created instance 1, we told all the other sleds about it, then when we create instance 2, there is no need for that creation saga to tell sled B where to find instance 1--we know that that was done already. So the instance creator doesn't have to do anything that will conflict with the migration, which prevents the race.

Right, this is what I was proposing (where any created instance already knows where previously created instances are), though now that I know that calling `set-v2p` overwrites previous mappings with new information, I'm not even sure `del-v2p` is needed!

I think this works out. Say the rule is:

> During instance creation or for the destination of a migration, add a mapping on every other sled: `new/migrated instance -> sled X`.

Let's take the same example as above: two instances, four sleds:

```
|A| |B| |C| |D|
|-| |-| |-| |-|
|1| |2| | | | |
| | | | | | | |
| | | | | | | |
```

Here's what would happen:

```
- OPTE automatically adds to sled A: instance 1 -> sled A
- add a mapping on sled B: instance 1 -> sled A
- add a mapping on sled C: instance 1 -> sled A
- add a mapping on sled D: instance 1 -> sled A

- add a mapping on sled A: instance 2 -> sled B
- OPTE automatically adds to sled B: instance 2 -> sled B
- add a mapping on sled C: instance 2 -> sled B
- add a mapping on sled D: instance 2 -> sled B
```

Now, instance 3 is created and allocated to sled C:

```
- add a mapping on sled A: instance 3 -> sled C
- add a mapping on sled B: instance 3 -> sled C
- OPTE automatically adds to sled C: instance 3 -> sled C
- add a mapping on sled D: instance 3 -> sled C
```

Now, say a concurrent migration is performed of instance 2 from B to D:

```
- add a mapping on sled A: instance 2 -> sled D
- add a mapping on sled B: instance 2 -> sled D
- add a mapping on sled C: instance 2 -> sled D
- OPTE automatically adds to sled D: instance 2 -> sled D
```

There's no mapping conflict here!

Reply:
We discussed this extensively offline and agreed that broadcasting mappings fixes these races, with one catch: we have to guarantee that we only program V2P mappings when the mapping programmer is certain that the instance's sled assignment is fixed. Fortunately:

- For instance create, this is trivial (the instance was just assigned a sled, and it's not running so it can't migrate).
- For migration it's a little more complicated (the migration-end logic needs to ensure that the "lock" preventing a second migration from starting is not "dropped" until the V2P mappings are re-established). But none of this logic exists yet (I'm working on it now, in fact), so not a problem for this PR.
- Eventually, we're going to want instance start to be a saga that selects a sled for the instance to run on. There we have to be careful not to hit races where e.g. an instance migrates, then *stops* (which Nexus generally can't prevent), then gets restarted on a different sled). But instance start isn't a saga today, and doesn't change any sled assignments, and so this is also not a problem for this PR.

In the fullness of time, it might be simpler just to have a generation number for V2P assignments as @bnaecker suggested [here](https://github.com/oxidecomputer/omicron/pull/2536#discussion_r1134943007) so that we don't have to worry so much about guaranteeing that there's only one assigner who's guaranteed to have the right assignments, but for now it seems like we might just get away with broadcasting mappings.

Homework assignments:

- @jmpesp will update the comment I had originally tagged to mention the "sled assignment can't change while we're doing this" invariant
- @gjcolombo will update RFD 361's discussion of what happens when migration finishes

(My thanks again to @jmpesp for letting me hash all this out here--the migration protocol is going to be better for having discussed all this in detail!)

Reply:
see e9c44118 for the updated comment. Thanks for the discussion Greg :)

Reply:
Ship it!

Reply:
@gjcolombo I added a comment in the instance migration saga about the unidirectional network blip we discussed in dbafcc12, have a look and let me know if I've missed anything

Reply:
New comment LGTM. 

If we're going to create the mappings in the saga, I would still probably prefer to put that it in its own saga step that precedes `sim_instance_migrate`, which would of course require the comment to be reworded slightly, but I think its essence is accurate.

Reply:
Sounds good, done in 9a731557